src/utils/hooks.ts

```diff
@@ -250,42 +250,61 @@ export function useFilters<T>(
+  const allFilterGroups = useMemo(() => {
+    return (options?.baseFilters ?? []).concat(filters ?? []);
+  }, [options?.baseFilters, filters]);
+
+  // Because searching through all components for matching groups/systems/etc is expensive
+  // we calculate one time and memoize for fast lookup.
+  const groupPropertyLookup: Record<string, Set<string>>[] | undefined = useMemo(() => {
+    return components === undefined
       ? undefined
       : allFilterGroups.map(filterGroup => {
-          return {
-            name: filterGroup.name,
-            filters: mapValues(
-              mapByString(
-                components.flatMap(
-                  component => filterGroup.groupProperty(component) ?? [],
-                ),
-                groupProperty => groupProperty,
+        return components
+          .reduce((mapSoFar, nextComponent) => {
+            filterGroup.groupProperty(nextComponent)?.forEach(groupProperty => {
+              if (mapSoFar[groupProperty] === undefined) {
+                mapSoFar[groupProperty] = new Set();
+              }
+              mapSoFar[groupProperty].add(stringifyAnyEntityRef(nextComponent))
+            })
+
+            return mapSoFar
+          }, {} as Record<string, Set<string>>);
+      })
+  }, [allFilterGroups, components]);
+
+  const filterGroups = useMemo(() => {
+    return components === undefined || groupPropertyLookup === undefined
+      ? undefined
+      : allFilterGroups.map((filterGroup, index) => {
+        return {
+          name: filterGroup.name,
+          filters: mapValues(
+            mapByString(
+              components.flatMap(
+                component => filterGroup.groupProperty(component) ?? [],
               ),
-              groupProperty => {
-                return {
-                  display: filterGroup.formatProperty
-                    ? filterGroup.formatProperty(groupProperty)
-                    : groupProperty,
-                  value: groupProperty,
-                };
-              },
+              groupProperty => groupProperty,
             ),
-            generatePredicate: (groupProperty: string) => {
-              return (t: T) => {
-                return components.some(
-                  component =>
-                    entityEquals(component, entityRef(t)) &&
-                    filterGroup
-                      .groupProperty(component)
-                      ?.includes(groupProperty),
-                );
+            groupProperty => {
+              return {
+                display: filterGroup.formatProperty
+                  ? filterGroup.formatProperty(groupProperty)
+                  : groupProperty,
+                value: groupProperty,
               };
             },
-          };
-        });
+          ),
+          generatePredicate: (groupProperty: string) => {
+            return (t: T) => {
+              return groupPropertyLookup[index][groupProperty].has(stringifyAnyEntityRef(entityRef(t)))
```

Comment:
Is this right? if `groupPropertyLookup` is `Record<string, Set<string>>` then I expect it to look something like
```json
{
  "key": ["item1", "item2"]
}
```
This looks like it's trying to access the Record with a number (which will likely return `undefined`) and then attempt to access a string property off of that.

Reply:
Oops it's late, I missed the list nature of `groupPropertyNature` 🙈. I think this is OK...
```ts
const groupPropertyLookup: Record<string, Set<string>>[]
```